### PR TITLE
Replace Hash-Based State with Value-Based State

### LINKED ISSUE #534: OD-7 NFTRenderer representing full properties
1. update NFTRenderer.VaultParams ✅ 
2. update NFTRenderer calls to pull new data ✅ 
3. update vault hash at `ODSafeManager.transferInternalCoins()` ✅ 
4. update: `coinBalance` and `tokenCollateral` cannot be added to the statehash. add direct check in transfer.

### LINKED ISSUE #561: Event emitter improvement
There are several contracts where setter functions do not emit an event:

- updateRedemptionRate function in OracleRelayer.sol contract.
- updateBlockDelay and updateTimeDelay functions in Vault721.sol contract.

### LINKED ISSUE #599: Vault721 does not inherit Modifiable.sol or Authorizable.sol
We should update Vault721.sol to inherit Authorizable and Modifiable so the dao can update params in the same way that other contracts are updated.

currently there is an onlyGovernance() modifier and some update functions (updateBlockDelay, updateTimeDelay etc).  I'd like to swap these out to update through the Modifiable.sol contract inheritances to keep it uniform with the rest of the contracts.

### LINKED ISSUE #600: Change StateHash to Uint256 with Value Checks
rather than hash collateral and debt, make sure collateral is greater than or equal, and debt is less than or equal, before and after token transfer.

### PR DESCRIPTION
**Pull request by daopunk**
closes #534 
closes #600 
closes #561
closes #599

Notes:
Due to the limitations of overriding the `tokenURI` function, specifically that it is a non-state changing view function that only returns a string, it is not possible to use it as a means of updating state information about NFVs. So the pattern of updating NFV states via applicable function calls from the ODSafeManager is maintained.

DOS attacks are reverted using a value-based state check that enforces static state or state that increases the value of the NFV after token transfer. Additionally, the logic to look up values in the SafeEngine has been added to the Vault721, so that it is not routing that call through the NFTRenderer contract, removing the NFTRenderer from state-enforcement checks.

Partially followed the recommendation; included `tokenCollateral` value, but not `coinBalance`. tokenCollateral was already in WAD in the SafeEngine, whereas coinBalance would need to be converted from RAD. These extra values are optional and only benefit the literate user with easier access to transparent data. tokenCollateral was chosen as the more relevant option to include in the json data.

### CODE REVIEW

**Comment by daopunk** [comment]
TokenCollateral, which should and will most likely be zero is the last attribute, positioned away from the Collateral attribute to avoid confusion.

Screenshot 2024-04-05 at 10 10 02 PM

**Comment by daopunk** [comment]
@pi0neerpat need to fix/change tests for this PR, but you can at least review the contract changes. I can write the tests tomorrow, before I forget.

File: src/contracts/proxies/Vault721.sol [hunk 1]
```diff
@@ -272,22 +283,32 @@ contract Vault721 is ERC721EnumerableUpgradeable {
...
+  /**
+   * @dev prevent undesirable frontrun state change during token transferFrom
+   * @notice frontrunning that increases the lockedCollateral or decreases the generatedDebt is accepted
    */
   function _enforceStaticState(address _operator, uint256 _tokenId) internal view {
-    // on allowlist addresses, we check the block delay along with the state hash
+    (uint256 _collateralNow, uint256 _debtNow) = _getNfvValue(_tokenId);
+    NFVState memory _nfv = _nfvState[_tokenId];
+
+    if (_collateralNow < _nfv.collateral || _debtNow > _nfv.debt) revert StateViolation();
```
Comment (pi0neerpat) [inline src/contracts/proxies/Vault721.sol:306]: Pending discussion

File: src/contracts/proxies/NFTRenderer.sol [hunk 2]
```diff
@@ -233,7 +213,9 @@ contract NFTRenderer {
       '"attributes":[{"trait_type":"ID","value":"',
       params.vaultId,
       traits,
-      params.lastUpdate
+      params.lastUpdate,
+      '"},{"trait_type":"TokenCollateral","value":"',
```
Comment (pi0neerpat) [inline src/contracts/proxies/NFTRenderer.sol:217]: I think we should rename to "Internal Collateral" to be even more explicit

**Review by pi0neerpat — CHANGES_REQUESTED** [review]

**Comment by daopunk** [comment]
Screenshot 2024-04-06 at 10 05 23 PM

**Comment by daopunk** [comment]
TokenURI is printing with all data, and non-formatted nums

**Comment by daopunk** [comment]
@MrDeadCe11 requesting help to finish some mocked unit tests in Vault721.t.sol that I commented out, since I am off.

**Comment by MrDeadCe11** [comment]
merged with pr #602 to include other updates to vault721, modify parameters and even emission improvements.

File: src/contracts/proxies/Vault721.sol (hunk 1 above)
Comment (MrDeadCe11) [reply src/contracts/proxies/Vault721.sol:193]: did this get resolved?

File: src/contracts/proxies/ODSafeManager.sol [hunk 3]
```diff
@@ -191,8 +195,7 @@ contract ODSafeManager is IODSafeManager, Authorizable, Modifiable {
       _sData.collateralType, _sData.safeHandler, collateralSource, debtDestination, _deltaCollateral, _deltaDebt
     );
 
-    vault721.updateVaultHashState(_safe);
-
+    vault721.updateNfvState(_safe);
```
Comment (pi0neerpat) [inline src/contracts/proxies/ODSafeManager.sol:198]: We may need to add a check here to only updateNfvState if the value is decreased, to avoid DoS

**Review by pi0neerpat — CHANGES_REQUESTED** [review]

**Review by pi0neerpat — APPROVED** [review]

**Pull request merged**